### ISSUE TITLE
Serial name of enum key in top-level map is ignored

### ISSUE THREAD

**Issue by xn32** [Kotlin/kotlinx.serialization#2225]
When encoding or decoding an enum-based map that is not wrapped within a (value) class, `SerialName` annotations seem to be ignored by the core library. The following code visualizes this issue for the JSON format:
```kotlin
@Serializable
enum class TestEnum {
    @SerialName("x") FIRST,
    @SerialName("y") SECOND
}

@JvmInline
@Serializable
value class Wrapper(val map: Map<TestEnum, Int>)

fun main() {
    println("Map in serializable class: " + Json.encodeToString(Wrapper(mapOf(TestEnum.FIRST to 0))))
    println("Map as top-level value: " + Json.encodeToString(mapOf(TestEnum.FIRST to 0)))
}
```

It produces the following output:
```
Map in serializable class: {"x":0}
Map as top-level value: {"FIRST":0}
```

Unless I am mistaken, `encodeToString` should return `{"x":0}` in both cases. I was able to reproduce this effect with other formats built upon the core library, i.e., it does not seem to be caused by the JSON format.

**Environment**
 - Kotlin version: 1.8.10
 - Library version: 1.5.0
 - Kotlin platforms: JVM
 - Gradle version: 7.4.2

**Comment by shanshin** [comment]
Duplicates #2183 and #2121.

Fixed in Kotlin 1.8.20

**Comment by sandwwraith** [comment]
@shanshin It looks like this is about annotation on members, not enum class itself. Maybe it's related to the intrinsic, which lack annotations at all currently.

**Comment by shanshin** [comment]
@sandwwraith, indeed.
But it's still fixed in `1.8.20` 
Instead of `new LinkedHashMapSerializer(EnumsKt.createSimpleEnumSerializer("TestEnum", (Enum[])TestEnum.values()), IntSerializer.INSTANCE))` generated `new LinkedHashMapSerializer(TestEnum.Companion.serializer(), IntSerializer.INSTANCE)`
